test_util/helpers.py

```diff
@@ -78,19 +78,45 @@ def api_request(self, method, path, host_url=None, **kwargs):
 
         request_url = path_join(host_url if host_url else self.default_host_url, path)
         headers.update(kwargs.pop('headers', {}))
-        logging.info('Request method {}: {}'.format(method, request_url))
+        logging.info('Request method {}: {}'.format(request_fn.__name__, request_url))
         logging.debug('Reqeust kwargs: {}'.format(kwargs))
         logging.debug('Request headers: {}'.format(headers))
-        return requests.request(method, request_url, headers=headers, **kwargs)
-
-    get = functools.partialmethod(api_request, 'get')
-    post = functools.partialmethod(api_request, 'post')
-    put = functools.partialmethod(api_request, 'put')
-    delete = functools.partialmethod(api_request, 'delete')
-    options = functools.partialmethod(api_request, 'options')
-    head = functools.partialmethod(api_request, 'head')
-    patch = functools.partialmethod(api_request, 'patch')
-    delete = functools.partialmethod(api_request, 'delete')
+        return request_fn(request_url, headers=headers, **kwargs)
+
+    get = functools.partialmethod(wrapper, requests.get)
+    post = functools.partialmethod(wrapper, requests.post)
+    put = functools.partialmethod(wrapper, requests.put)
+    delete = functools.partialmethod(wrapper, requests.delete)
+    options = functools.partialmethod(wrapper, requests.options)
+    head = functools.partialmethod(wrapper, requests.head)
+    patch = functools.partialmethod(wrapper, requests.patch)
+    delete = functools.partialmethod(wrapper, requests.delete)
+
+
+class NodeApiClient(ApiClient):
+    def __init__(self, get_node_url, *args, **kwargs):
+        """ ApiClient variant allowing node awareness into the requests wrapper
+        """
+        self._get_node_url = get_node_url
+        super().__init__(*args, **kwargs)
+
+    def wrapper(self, request_fn, path, host_url=None, node=None, port=None, **kwargs):
+        """If node kwarg is given, a URL will be fetched from _get_node_url instead
+        of using the default host url from ApiClient
+        """
+        if node is not None:
+            assert host_url is not None, 'Cannot set both node and host_url'
+            host_url = self._get_node_url(node, port)
+        super().wrapper(request_fn, path, host_url=host_url, **kwargs)
+
+    get = functools.partialmethod(wrapper, requests.get)
```

Comment:
When we declare methods this way instead of doing the hackish setattr method, we have to copy/paste this into each class that wants to touch 'wrapper'